supportedBackends/sfs/src/main/scala/cromwell/backend/sfs/SharedFileSystemAsyncJobExecutionActor.scala

```diff
@@ -266,6 +291,44 @@ trait SharedFileSystemAsyncJobExecutionActor
           |""".stripMargin)
   }
 
+  /**
+    * Send the job id of the running job to the key value store.
+    *
+    * @param runningJobOption The running job.
+    */
+  private def tellKvJobId(runningJobOption: SharedFileSystemJob): Unit = {
+    val kvJobKey =
+      KvJobKey(jobDescriptor.key.call.fullyQualifiedName, jobDescriptor.key.index, jobDescriptor.key.attempt)
+    val scopedKey = ScopedKey(jobDescriptor.workflowDescriptor.id, kvJobKey, SharedFileSystemJob.JobIdKey)
+    val kvValue = Option(runningJobOption.jobId)
+    val kvPair = KvPair(scopedKey, kvValue)
+    val kvPut = KvPut(kvPair)
+    serviceRegistryActor ! kvPut
+  }
+
+  def recoverScript(job: SharedFileSystemJob): ExecutionHandle = {
+    context.become(sharedReceive(Option(job)) orElse super.receive)
+    // To avoid race conditions, check for the rc file after checking if the job is alive.
+    if (isAlive(job) || jobPaths.returnCode.exists) {
```

Comment:
The FS check is absolutely easier. However, there exists a brief time span where we might run into problems if we check in opposite order of `jobPaths.returnCode.exists || isAlive(job)`:

cromwell | job
---|---
check rc file exists (false) | ...
... | write rc file
... | job ends
check job is alive (false) | ...

If we reverse the order of the checks, we'll never<sup>1</sup> run into the above race condition.

---

<sup>1</sup> The network file system (NFS) is a fickle beast, and past pipeline managers have seen files "suddenly" appear after some latency. It is possible that the job may finish, and the RC file may not be visible to cromwell. I'm taking a "wait and see" approach to seeing if this issue actually appears.